eslint.config.js

```diff
@@ -0,0 +1,245 @@
+    },
+    rules: {
+      "prettier/prettier": "error",
+      "import/no-extraneous-dependencies": "error",
+    },
+  },
+  {
+    files: ["**/*.ts"],
+    languageOptions: {
+      parser: parserTypeScriptESLint,
+      parserOptions: {
+        project: "./tsconfig.json",
+      },
+    },
+    plugins: {
+      "@typescript-eslint": pluginTypeScriptESLint,
+    },
+    rules: {
+      "no-unused-vars": "off",
+      "@typescript-eslint/no-unused-vars": "error",
+      "no-dupe-class-members": "off",
+      "@typescript-eslint/no-dupe-class-members": "error",
+      "no-undef": "off",
+      "no-redeclare": "off",
+      "@typescript-eslint/ban-ts-comment": [
+        "error",
+        {
+          "ts-ignore": { descriptionFormat: "^\\(Babel 7 vs Babel 8\\) .+$" },
+        },
+      ],
+      "@typescript-eslint/consistent-type-imports": [
+        "error",
+        { disallowTypeAnnotations: false },
+      ],
+      "@typescript-eslint/no-use-before-define": [
+        "error",
+        {
+          functions: false,
+          classes: false,
+          allowNamedExports: true,
+        },
+      ],
+      "@typescript-eslint/no-unnecessary-type-assertion": "error",
+      "@typescript-eslint/no-confusing-void-expression": [
+        "error",
+        { ignoreArrowShorthand: true },
+      ],
+      "@typescript-eslint/no-import-type-side-effects": "error",
+    },
+  },
+  {
+    files: sourceFiles("js,ts,cjs,mjs"),
+    languageOptions: {
+      globals: { PACKAGE_JSON: "readonly", USE_ESM: "readonly" },
+    },
+    rules: {
+      "@babel/development/no-undefined-identifier": "error",
+      "@babel/development/no-deprecated-clone": "error",
+      "guard-for-in": "error",
+      "import/extensions": ["error", { json: "always", cjs: "always" }],
+    },
+  },
+  ...compat.extends("plugin:jest/recommended").map(config => {
+    config.files = [
+      ...testFiles,
+      "packages/babel-helper-transform-fixture-test-runner/src/helpers.{ts,js}",
+      "test/**/*.js",
+    ];
+    config.rules = {
+      ...config.rules,
+      "jest/expect-expect": "off",
+      "jest/no-identical-title": "off",
+      "jest/no-standalone-expect": "off",
+      "jest/no-test-callback": "off",
+      "jest/valid-describe": "off",
+      "import/extensions": ["error", "always"],
+      "import/no-extraneous-dependencies": "off",
+      "no-restricted-imports": ["error", { patterns: ["**/src/**"] }],
+    };
+    return config;
+  }),
+  {
+    files: testFiles,
+    rules: {
+      "node/no-unsupported-features": [
+        "error",
+        { version: "12.17.0", ignores: ["modules"] },
+      ],
+      "@babel/development-internal/require-default-import-fallback": "error",
+    },
+  },
+  {
+    files: [...sourceFiles("js,ts,mjs"), ...testFiles, "test/**/*.js"],
+    ignores: [
+      // @babel/register is the require() hook, so it will always be CJS-based
+      "packages/babel-register/**/*.{js,ts}",
+    ],
```

Comment:
`excludedFiles` is replaced by `ignores` because the former is not supported in flat config.
If I understand correctly, `excludedFiles` will exclude the matched file from this override config item but ESLint will continue matching it with other config items, while `ignores` will stop ESLint matching further config items to this file. @nzakas Is that correct?

Here it can be replaced because there are no more rules matching `packages/babel-register/**/*.{js,ts}`.

Reply:
Almost. When both `files` and `ignores` are present in the same object, `ignores` acts like `excludedFiles` in that it simply filters out results that match `files`.

When `ignores` is present with any other key (i.e. `rules`) then it also acts like `excludedFiles` in that it matches every file except the ones matching patterns in `ignores`.

When `ignores` is present without any other keys in the same object then those files will always be ignored and never match any config objects.

Reply:
Thank you! That is very helpful. I can't find much docs on `ignores` on the website, should I open a PR to update https://eslint.org/docs/latest/use/configure/ignore?

Reply:
It's on this page:
https://eslint.org/docs/latest/use/configure/configuration-files-new#specifying-files-and-ignores

Once we officially make flat config the default, we'll update the URL so this one is the default and the eslintrc one is marked as deprecated.

Reply:
Oh thank you! Now I might understand why searching "ignores" on the website does not return the options docs: most of them are rules options.

Are they supposed to show up once the flat config is made default?

Reply:
Yes. When we release v9.0.0, we will rearrange the documentation so that we favor everything related to flat config and the eslintrc docs will be made less prominent. So when you look up "ignores" it will show you docs related to flat config.

We're just in this weird intermediate state right now.